### ISSUE TITLE
Cannot get PHP working

### ISSUE THREAD

**Issue by leesei** [caddyserver/caddy#4181]
Caddy version:  v2.4.1 h1:kAJ0JB5Xk5gPdTH/27S5cyoMGqD5lBAe9yZ8zTjVJa0=

Caddyfile:

```
http://localhost:8888 {

    log {
        level debug
        output file /var/log/caddy/access.log
        # format single_field common_log
    }

    root * /home/leesei/wip/caddy-php

    php_fastcgi unix//run/php-fpm7/php-fpm.sock {
        index phpinfo.php
        resolve_root_symlink
    }

    encode gzip
    file_server
}
```

This is my directory structure:

```sh
Caddyfile  README.md  phpinfo.php
```

---

http://localhost:8888/ returns 

```
2021/05/28 10:17:26.649	error	http.log.access.log0	handled request	{"request": {"remote_addr": "[::1]:41360", "proto": "HTTP/1.1", "method": "GET", "host": "localhost:8888", "uri": "/", "headers": {"Connection": ["keep-alive"], "Sec-Ch-Ua-Mobile": ["?0"], "Accept-Encoding": ["gzip, deflate, br"], "User-Agent": ["Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.85 Safari/537.36"], "Sec-Fetch-Site": ["none"], "Sec-Fetch-Dest": ["document"], "Accept-Language": ["en-US,en-HK;q=0.9,en;q=0.8,zh-HK;q=0.7,zh;q=0.6,zh-TW;q=0.5,zh-CN;q=0.4,ja;q=0.3,es;q=0.2"], "Upgrade-Insecure-Requests": ["1"], "Accept": ["text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9"], "Sec-Fetch-User": ["?1"], "Cache-Control": ["max-age=0"], "Sec-Ch-Ua": ["\" Not A;Brand\";v=\"99\", \"Chromium\";v=\"90\", \"Google Chrome\";v=\"90\""], "Dnt": ["1"], "Sec-Fetch-Mode": ["navigate"]}}, "common_log": "::1 - - [28/May/2021:18:17:26 +0800] \"GET / HTTP/1.1\" 404 0", "duration": 0.000291479, "size": 0, "status": 404, "resp_headers": {"Server": ["Caddy"]}}
```

http://localhost:8888/phpinfo.php returns 

```
2021/05/28 10:18:05.168	error	http.log.access.log0	handled request	{"request": {"remote_addr": "[::1]:41702", "proto": "HTTP/1.1", "method": "GET", "host": "localhost:8888", "uri": "/phpinfo.php", "headers": {"Connection": ["keep-alive"], "Sec-Ch-Ua": ["\" Not A;Brand\";v=\"99\", \"Chromium\";v=\"90\", \"Google Chrome\";v=\"90\""], "Sec-Fetch-User": ["?1"], "Sec-Ch-Ua-Mobile": ["?0"], "Sec-Fetch-Dest": ["document"], "Dnt": ["1"], "User-Agent": ["Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.85 Safari/537.36"], "Sec-Fetch-Site": ["none"], "Accept-Language": ["en-US,en-HK;q=0.9,en;q=0.8,zh-HK;q=0.7,zh;q=0.6,zh-TW;q=0.5,zh-CN;q=0.4,ja;q=0.3,es;q=0.2"], "Upgrade-Insecure-Requests": ["1"], "Accept": ["text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9"], "Sec-Fetch-Mode": ["navigate"], "Accept-Encoding": ["gzip, deflate, br"]}}, "common_log": "::1 - - [28/May/2021:18:18:05 +0800] \"GET /phpinfo.php HTTP/1.1\" 404 16", "duration": 0.000592642, "size": 16, "status": 404, "resp_headers": {"X-Powered-By": ["PHP/7.4.18"], "Content-Type": ["text/html; charset=UTF-8"], "Server": ["Caddy"], "Status": ["404 Not Found"]}}
```

My PHP-FPM log:

```
- -  28/May/2021:18:18:05 +0800 [-] "GET /phpinfo.php" 404 - 0.116 2048 0.00%
```

`[-]` after time is `%f: script filename`, I wonder if the root for CGI is setup incorrectly.

**Comment by Mohammed90** [comment]
Thanks for your question, and we're thrilled that you're using Caddy! This looks more like a question about how to use Caddy rather than a bug report or feature request. Since this issue tracker is reserved for actionable development items, I'm going to close this, but we have a [community forum](https://caddy.community) where more people will be exposed to your question, including people who may be more expert or experienced with the specific question you're facing. I hope you'll ask your question there, and thanks for understanding!

**Comment by ScrumpyJack** [comment]
I have the same issue but I can't get to caddy.community
@leesei How did you fix this?

**Comment by leesei** [comment]
> https://caddy.community/t/cannot-get-php-working-in-caddy-v2/12637/5

Following the user direction, I’ve solved the issue.

1. `php-fpm` is started as user http
the socket is also owned by http and has 0600 permission
so `sudo -u http caddy start` is enough
2. php-fpm requires access of all folders all the way to www root
I probably enabled error log in PHP to see this log and I saw it failed as the http user tries to `lstat /home/leesei/wip/`
this is where I got stuck with, I tried `chmod -r http:http /home/leesei/wip/learn-php` already

Solution:

- moved my www root to `/srv/http/learn-php` (`/srv/http` is owned by http user)
- update Caddyfile
- sudo -u http caddy start
- have fun with PHP

---

Use `cgi-fcgi` for testing your PHP works first.
http://www.inanzzz.com/index.php/post/653v/testing-php-fpm-without-having-a-web-server
https://easyengine.io/tutorials/php/directly-connect-php-fpm/
